File: app/src/main/java/com/vteam/testdemo/chat/adapter/MessageAdapter.java
package com.vteam.testdemo.chat.adapter;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.vteam.testdemo.R;
import com.vteam.testdemo.landing.LandingActivity;
import com.vteam.testdemo.top.ImageViewerActivity;
import com.vteam.testdemo.top.Messages;

import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;

public class MessageAdapter extends RecyclerView.Adapter<MessageAdapter.MessageViewHolder> {
    private List<Messages> userMessagesList;
    private FirebaseAuth mAuth;
    private DatabaseReference usersRef;


    public MessageAdapter(List<Messages> userMessagesList) {
        this.userMessagesList = userMessagesList;
    }

    @NonNull
    @Override
    public MessageViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.custom_messages_layout, viewGroup, false);

        mAuth = FirebaseAuth.getInstance();

        return new MessageViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final MessageViewHolder messageViewHolder, final int position) {
        String messageSenderId = mAuth.getCurrentUser().getUid();
        Messages messages = userMessagesList.get(position);

        String fromUserID = messages.getFrom();
        String fromMessageType = messages.getType();

        usersRef = FirebaseDatabase.getInstance().getReference().child("Users").child(fromUserID);

        usersRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                if (dataSnapshot.hasChild("image")) {
                    String receiverImage = dataSnapshot.child("image").getValue().toString();

                    if (!receiverImage.isEmpty()) {
                        StorageReference reference = FirebaseStorage.getInstance().getReference();
                        reference.child(receiverImage).getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                            @Override
                            public void onSuccess(Uri uri) {
                                // Got the download URL for 'users/me/profile.png'
                                Log.d("URL", "" + uri);


                                Glide.with(messageViewHolder.receiverProfileImage.getContext())
                                        .load(uri)
                                        .into(messageViewHolder.receiverProfileImage);
                            }
                        }).addOnFailureListener(new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception exception) {
                                // Handle any errors
                            }
                        });
                    }
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });


        messageViewHolder.receiverMessageText.setVisibility(View.GONE);
        messageViewHolder.receiverProfileImage.setVisibility(View.GONE);
        messageViewHolder.senderMessageText.setVisibility(View.GONE);
        messageViewHolder.messageSenderPicture.setVisibility(View.GONE);
        messageViewHolder.messageReceiverPicture.setVisibility(View.GONE);


        if (fromMessageType.equals("text")) {
            if (fromUserID.equals(messageSenderId)) {
                messageViewHolder.senderMessageText.setVisibility(View.VISIBLE);

                messageViewHolder.senderMessageText.setBackgroundResource(R.drawable.sender_messages_layout);
                messageViewHolder.senderMessageText.setTextColor(Color.BLACK);
                messageViewHolder.senderMessageText.setText(messages.getMessage() + "\n \n" + messages.getTime() + " - " + messages.getDate());
            } else {
                messageViewHolder.receiverProfileImage.setVisibility(View.VISIBLE);
                messageViewHolder.receiverMessageText.setVisibility(View.VISIBLE);

                messageViewHolder.receiverMessageText.setBackgroundResource(R.drawable.receiver_messages_layout);
                messageViewHolder.receiverMessageText.setTextColor(Color.BLACK);
                messageViewHolder.receiverMessageText.setText(messages.getMessage() + "\n \n" + messages.getTime() + " - " + messages.getDate());
            }
        } else if (fromMessageType.equals("image")) {
            if (fromUserID.equals(messageSenderId)) {
                messageViewHolder.messageSenderPicture.setVisibility(View.VISIBLE);

                if (!messages.getMessage().isEmpty()) {
                    StorageReference reference = FirebaseStorage.getInstance().getReference();
                    reference.child(messages.getMessage()).getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                        @Override
                        public void onSuccess(Uri uri) {
                            // Got the download URL for 'users/me/profile.png'
                            Log.d("URL", "" + uri);


                            Glide.with(messageViewHolder.messageSenderPicture.getContext())
                                    .load(uri)
                                    .into(messageViewHolder.messageSenderPicture);
                        }
                    }).addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception exception) {
                            // Handle any errors
                        }
                    });
                }

            } else {
                messageViewHolder.receiverProfileImage.setVisibility(View.VISIBLE);
                messageViewHolder.messageReceiverPicture.setVisibility(View.VISIBLE);
                if (!messages.getMessage().isEmpty()) {
                    StorageReference reference = FirebaseStorage.getInstance().getReference();
                    reference.child(messages.getMessage()).getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
                        @Override
                        public void onSuccess(Uri uri) {
                            // Got the download URL for 'users/me/profile.png'
                            Log.d("URL", "" + uri);


                            Glide.with(messageViewHolder.messageReceiverPicture.getContext())
                                    .load(uri)
                                    .into(messageViewHolder.messageReceiverPicture);
                        }
                    }).addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception exception) {
                            // Handle any errors
                        }
                    });
                }

            }
        } else if (fromMessageType.equals("pdf") || fromMessageType.equals("docx")) {
            if (fromUserID.equals(messageSenderId)) {
                messageViewHolder.messageSenderPicture.setVisibility(View.VISIBLE);
//                messageViewHolder.messageSenderPicture.setBackground(R.drawable.);

                messageViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(userMessagesList.get(position).getMessage()));
                        messageViewHolder.itemView.getContext().startActivity(intent);
                    }
                });
            } else {
                messageViewHolder.receiverProfileImage.setVisibility(View.VISIBLE);
                messageViewHolder.messageReceiverPicture.setVisibility(View.VISIBLE);
//                messageViewHolder.messageReceiverPicture.setBackground(R.drawable.file);
            }
        }
        if (fromUserID.equals(messageSenderId)) {
            messageViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (userMessagesList.get(position).getType().equals("pdf") || userMessagesList.get(position).getType().equals("docx")) {
                        CharSequence options[] = new CharSequence[]
                                {
                                        "Delete For me",
                                        "Download and View this Document",
                                        "Cancel",
                                        "Delete For EveryOne"
                                };
                        AlertDialog.Builder builder = new AlertDialog.Builder(messageViewHolder.itemView.getContext());
                        builder.setTitle("Delete Message");

                        builder.setItems(options, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int position) {
                                if (position == 0) {
                                    deleteSentMessage(position, messageViewHolder);
                                    Intent intent = new Intent(messageViewHolder.itemView.getContext(), LandingActivity.class);
                                    messageViewHolder.itemView.getContext().startActivity(intent);

                                } else if (position == 1) {
                                    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(userMessagesList.get(position).getMessage()));
                                    messageViewHolder.itemView.getContext().startActivity(intent);
                                } else if (position == 3) {
                                    deleteMessageForEveryOne(position, messageViewHolder);
                                }
                            }
                        });
                        builder.show();
                    } else if (userMessagesList.get(position).getType().equals("text")) {
                        CharSequence options[] = new CharSequence[]
                                {
                                        "Delete For me",
                                        "Cancel",
                                        "Delete For EveryOne"
                                };
                        AlertDialog.Builder builder = new AlertDialog.Builder(messageViewHolder.itemView.getContext());
                        builder.setTitle("Delete Message");

                        builder.setItems(options, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int position) {
                                if (position == 0) {
                                    deleteSentMessage(position, messageViewHolder);
                                    Intent intent = new Intent(messageViewHolder.itemView.getContext(), LandingActivity.class);
                                    messageViewHolder.itemView.getContext().startActivity(intent);

                                } else if (position == 2) {
                                    deleteMessageForEveryOne(position, messageViewHolder);
                                }
                            }
                        });
                        builder.show();
                    } else if (userMessagesList.get(position).getType().equals("image")) {
                        CharSequence options[] = new CharSequence[]
                                {
                                        "Delete For me",
                                        "View This Image",
                                        "Cancel",
                                        "Delete For EveryOne"
                                };
                        AlertDialog.Builder builder = new AlertDialog.Builder(messageViewHolder.itemView.getContext());
                        builder.setTitle("Delete Message");

                        builder.setItems(options, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int position) {
                                if (position == 0) {
                                    deleteSentMessage(position, messageViewHolder);
                                    Intent intent = new Intent(messageViewHolder.itemView.getContext(), LandingActivity.class);
                                    messageViewHolder.itemView.getContext().startActivity(intent);

                                } else if (position == 1) {
                                    Intent intent = new Intent(messageViewHolder.itemView.getContext(), ImageViewerActivity.class);
                                    intent.putExtra("url", userMessagesList.get(position).getMessage());
                                    messageViewHolder.itemView.getContext().startActivity(intent);
                                } else if (position == 3) {
                                    deleteMessageForEveryOne(position, messageViewHolder);
                                }
                            }
                        });
                        builder.show();
                    }
                }
            });
        } else {
            messageViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (userMessagesList.get(position).getType().equals("pdf") || userMessagesList.get(position).getType().equals("docx")) {
                        CharSequence options[] = new CharSequence[]
                                {
                                        "Delete For me",
                                        "Download and View this Document",
                                        "Cancel"
                                };
                        AlertDialog.Builder builder = new AlertDialog.Builder(messageViewHolder.itemView.getContext());
                        builder.setTitle("Delete Message");

                        builder.setItems(options, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int position) {
                                if (position == 0) {
                                    deleteReceiverMessage(position, messageViewHolder);
                                    Intent intent = new Intent(messageViewHolder.itemView.getContext(), LandingActivity.class);
                                    messageViewHolder.itemView.getContext().startActivity(intent);

                                } else if (position == 1) {
                                    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(userMessagesList.get(position).getMessage()));
                                    messageViewHolder.itemView.getContext().startActivity(intent);
                                }
                            }
                        });
                        builder.show();
                    } else if (userMessagesList.get(position).getType().equals("text")) {
                        CharSequence options[] = new CharSequence[]
                                {
                                        "Delete For me",
                                        "Cancel"
                                };
                        AlertDialog.Builder builder = new AlertDialog.Builder(messageViewHolder.itemView.getContext());
                        builder.setTitle("Delete Message");

                        builder.setItems(options, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int position) {
                                if (position == 0) {
                                    deleteReceiverMessage(position, messageViewHolder);
                                    Intent intent = new Intent(messageViewHolder.itemView.getContext(), LandingActivity.class);
                                    messageViewHolder.itemView.getContext().startActivity(intent);
                                }
                            }
                        });
                        builder.show();
                    } else if (userMessagesList.get(position).getType().equals("image")) {
                        CharSequence options[] = new CharSequence[]
                                {
                                        "Delete For me",
                                        "View This Image",
                                        "Cancel"
                                };
                        AlertDialog.Builder builder = new AlertDialog.Builder(messageViewHolder.itemView.getContext());
                        builder.setTitle("Delete Message");

                        builder.setItems(options, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int position) {
                                if (position == 0) {
                                    deleteReceiverMessage(position, messageViewHolder);
                                    Intent intent = new Intent(messageViewHolder.itemView.getContext(), LandingActivity.class);
                                    messageViewHolder.itemView.getContext().startActivity(intent);

                                } else if (position == 1) {
                                    Intent intent = new Intent(messageViewHolder.itemView.getContext(), ImageViewerActivity.class);
                                    intent.putExtra("url", userMessagesList.get(position).getMessage());
                                    messageViewHolder.itemView.getContext().startActivity(intent);
                                }
                            }
                        });
                        builder.show();
                    }
                }
            });

        }
    }

    @Override
    public int getItemCount() {
        return userMessagesList.size();
    }

    private void deleteSentMessage(final int position, final MessageViewHolder holder) {
        DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
        rootRef.child("Messages")
                .child(userMessagesList.get(position).getFrom())
                .child(userMessagesList.get(position).getTo())
                .child(userMessagesList.get(position).getMessageID())
                .removeValue().addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    Toast.makeText(holder.itemView.getContext(), "Deleted Successfully.", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(holder.itemView.getContext(), "Error Occurred.", Toast.LENGTH_SHORT).show();
                }
            }
        });

    }

    private void deleteReceiverMessage(final int position, final MessageViewHolder holder) {
        DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
        rootRef.child("Messages")
                .child(userMessagesList.get(position).getTo())
                .child(userMessagesList.get(position).getFrom())
                .child(userMessagesList.get(position).getMessageID())
                .removeValue().addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    Toast.makeText(holder.itemView.getContext(), "Deleted Successfully.", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(holder.itemView.getContext(), "Error Occurred.", Toast.LENGTH_SHORT).show();
                }
            }
        });

    }

    private void deleteMessageForEveryOne(final int position, final MessageViewHolder holder) {
        final DatabaseReference rootRef = FirebaseDatabase.getInstance().getReference();
        rootRef.child("Messages")
                .child(userMessagesList.get(position).getFrom())
                .child(userMessagesList.get(position).getTo())
                .child(userMessagesList.get(position).getMessageID())
                .removeValue().addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    rootRef.child("Messages")
                            .child(userMessagesList.get(position).getFrom())
                            .child(userMessagesList.get(position).getTo())
                            .child(userMessagesList.get(position).getMessageID())
                            .removeValue().addOnCompleteListener(new OnCompleteListener<Void>() {
                        @Override
                        public void onComplete(@NonNull Task<Void> task) {
                            if (task.isSuccessful()) {
                                Toast.makeText(holder.itemView.getContext(), "Error Occurred.", Toast.LENGTH_SHORT).show();
                            }
                        }
                    });
                } else {
                    Toast.makeText(holder.itemView.getContext(), "Error Occurred.", Toast.LENGTH_SHORT).show();
                }
            }
        });

    }

    public class MessageViewHolder extends RecyclerView.ViewHolder {
        public TextView senderMessageText, receiverMessageText;
        public CircleImageView receiverProfileImage;
        public ImageView messageSenderPicture, messageReceiverPicture;


        public MessageViewHolder(@NonNull View itemView) {
            super(itemView);

            senderMessageText = (TextView) itemView.findViewById(R.id.sender_message_text);
            receiverMessageText = (TextView) itemView.findViewById(R.id.receiver_message_text);
            receiverProfileImage = (CircleImageView) itemView.findViewById(R.id.message_profile_image);
            messageReceiverPicture = itemView.findViewById(R.id.message_receiver_image_view);
            messageSenderPicture = itemView.findViewById(R.id.message_sender_image_view);
        }
    }
}

File: app/src/main/java/com/vteam/testdemo/chat/adapter/MessageAdapter.kt
package com.vteam.testdemo.chat.adapter

import android.app.AlertDialog
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage
import com.vteam.testdemo.R
import com.vteam.testdemo.chat.adapter.MessageAdapter.MessageViewHolder
import com.vteam.testdemo.common.ConstantNodes
import com.vteam.testdemo.databinding.CustomMessagesLayoutBinding
import com.vteam.testdemo.landing.LandingActivity
import com.vteam.testdemo.top.ImageViewerActivity
import com.vteam.testdemo.top.Messages

class MessageAdapter(private val userMessagesList: List<Messages>) : RecyclerView.Adapter<MessageViewHolder>() {

    private var auth: FirebaseAuth? = null
    private var usersRef: DatabaseReference? = null



    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): MessageViewHolder {
        val binding: CustomMessagesLayoutBinding = DataBindingUtil.inflate(
            LayoutInflater.from(viewGroup.context),
            R.layout.custom_messages_layout,
            viewGroup,
            false
        )

        auth = FirebaseAuth.getInstance()
        return MessageViewHolder(binding)
    }

    override fun onBindViewHolder(
        messageViewHolder: MessageViewHolder,
        position: Int
    ) {
        val messageSenderId = auth!!.currentUser!!.uid
        val messages = userMessagesList[position]
        val fromUserID = messages.from
        val fromMessageType = messages.type


        usersRef = FirebaseDatabase.getInstance().reference.child(ConstantNodes.NODES.USER_NODE).child(fromUserID)
        usersRef!!.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.hasChild("image")) {
                    val receiverImage =
                        dataSnapshot.child("image").value.toString()
                    if (receiverImage.isNotEmpty()) {
                        val reference =
                            FirebaseStorage.getInstance().reference
                        reference.child(receiverImage).downloadUrl
                            .addOnSuccessListener { uri -> // Got the download URL for 'users/me/profile.png'
                                Log.d("URL", "" + uri)
                                Glide.with(messageViewHolder.binding.messageReceiverImageView.context)
                                    .load(uri)
                                    .into(messageViewHolder.binding.messageReceiverImageView)
                            }.addOnFailureListener {
                                // Handle any errors
                            }
                    }
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })

        messageViewHolder.binding.receiverMessageText.visibility = View.GONE
        messageViewHolder.binding.messageReceiverImageView.visibility = View.GONE
        messageViewHolder.binding.senderMessageText.visibility = View.GONE
        messageViewHolder.binding.messageSenderImageView.visibility = View.GONE
        messageViewHolder.binding.messageReceiverImageView.visibility = View.GONE
        if (fromMessageType == "text") {
            if (fromUserID == messageSenderId) {
                messageViewHolder.binding.senderMessageText.visibility = View.VISIBLE
                messageViewHolder.binding.senderMessageText.setBackgroundResource(R.drawable.receiver_messages_layout)
                messageViewHolder.binding.senderMessageText.setTextColor(Color.BLACK)
                messageViewHolder.binding.senderMessageText.text = """${messages.message}

${messages.time} - ${messages.date}"""
            } else {
                messageViewHolder.binding.messageReceiverImageView.visibility = View.VISIBLE
                messageViewHolder.binding.receiverMessageText.visibility = View.VISIBLE
                messageViewHolder.binding.receiverMessageText.setBackgroundResource(R.drawable.receiver_messages_layout)
                messageViewHolder.binding.receiverMessageText.setTextColor(Color.BLACK)
                messageViewHolder.binding.receiverMessageText.text = """${messages.message}

${messages.time} - ${messages.date}"""
            }
        } else if (fromMessageType == "image") {
            if (fromUserID == messageSenderId) {
                messageViewHolder.binding.messageSenderImageView.visibility = View.VISIBLE
                if (messages.message.isNotEmpty()) {
                    val reference = FirebaseStorage.getInstance().reference
                    reference.child(messages.message).downloadUrl
                        .addOnSuccessListener { uri -> // Got the download URL for 'users/me/profile.png'
                            Log.d("URL", "" + uri)
                            Glide.with(messageViewHolder.binding.messageSenderImageView.context)
                                .load(uri)
                                .into(messageViewHolder.binding.messageSenderImageView)
                        }.addOnFailureListener {
                            // Handle any errors
                        }
                }
            } else {
                messageViewHolder.binding.messageReceiverImageView.visibility = View.VISIBLE
                messageViewHolder.binding.messageReceiverImageView.visibility = View.VISIBLE
                if (messages.message.isNotEmpty()) {
                    val reference = FirebaseStorage.getInstance().reference
                    reference.child(messages.message).downloadUrl
                        .addOnSuccessListener { uri -> // Got the download URL for 'users/me/profile.png'
                            Log.d("URL", "" + uri)
                            Glide.with(messageViewHolder.binding.messageReceiverImageView.context)
                                .load(uri)
                                .into(messageViewHolder.binding.messageReceiverImageView)
                        }.addOnFailureListener {
                            // Handle any errors
                        }
                }
            }
        } else if (fromMessageType == "pdf" || fromMessageType == "docx") {
            if (fromUserID == messageSenderId) {
                messageViewHolder.binding.messageSenderImageView.visibility = View.VISIBLE
                //                messageViewHolder.messageSenderPicture.setBackground(R.drawable.);
                messageViewHolder.itemView.setOnClickListener {
                    val intent = Intent(
                        Intent.ACTION_VIEW,
                        Uri.parse(userMessagesList[position].message)
                    )
                    messageViewHolder.itemView.context.startActivity(intent)
                }
            } else {
                messageViewHolder.binding.messageReceiverImageView.visibility = View.VISIBLE
                messageViewHolder.binding.messageReceiverImageView.visibility = View.VISIBLE
                //                messageViewHolder.messageReceiverPicture.setBackground(R.drawable.file);
            }
        }
        if (fromUserID == messageSenderId) {
            messageViewHolder.itemView.setOnClickListener {
                if (userMessagesList[position]
                        .type == "pdf" || userMessagesList[position]
                        .type == "docx"
                ) {
                    val options =
                        arrayOf<CharSequence>(
                            "Delete For me",
                            "Download and View this Document",
                            "Cancel",
                            "Delete For EveryOne"
                        )
                    val builder =
                        AlertDialog.Builder(messageViewHolder.itemView.context)
                    builder.setTitle("Delete Message")
                    builder.setItems(options) { dialogInterface, position ->
                        when (position) {
                            0 -> {
                                deleteSentMessage(position, messageViewHolder)
                                val intent = Intent(
                                    messageViewHolder.itemView.context,
                                    LandingActivity::class.java
                                )
                                messageViewHolder.itemView.context.startActivity(intent)
                            }
                            1 -> {
                                val intent = Intent(
                                    Intent.ACTION_VIEW,
                                    Uri.parse(
                                        userMessagesList[position].message
                                    )
                                )
                                messageViewHolder.itemView.context.startActivity(intent)
                            }
                            3 -> {
                                deleteMessageForEveryOne(position, messageViewHolder)
                            }
                        }
                    }
                    builder.show()
                } else if (userMessagesList[position].type == "text") {
                    val options =
                        arrayOf<CharSequence>(
                            "Delete For me",
                            "Cancel",
                            "Delete For EveryOne"
                        )
                    val builder =
                        AlertDialog.Builder(messageViewHolder.itemView.context)
                    builder.setTitle("Delete Message")
                    builder.setItems(options) { dialogInterface, position ->
                        if (position == 0) {
                            deleteSentMessage(position, messageViewHolder)
                            val intent = Intent(
                                messageViewHolder.itemView.context,
                                LandingActivity::class.java
                            )
                            messageViewHolder.itemView.context.startActivity(intent)
                        } else if (position == 2) {
                            deleteMessageForEveryOne(position, messageViewHolder)
                        }
                    }
                    builder.show()
                } else if (userMessagesList[position].type == "image") {
                    val options =
                        arrayOf<CharSequence>(
                            "Delete For me",
                            "View This Image",
                            "Cancel",
                            "Delete For EveryOne"
                        )
                    val builder =
                        AlertDialog.Builder(messageViewHolder.itemView.context)
                    builder.setTitle("Delete Message")
                    builder.setItems(options) { dialogInterface, position ->
                        when (position) {
                            0 -> {
                                deleteSentMessage(position, messageViewHolder)
                                val intent = Intent(
                                    messageViewHolder.itemView.context,
                                    LandingActivity::class.java
                                )
                                messageViewHolder.itemView.context.startActivity(intent)
                            }
                            1 -> {
                                val intent = Intent(
                                    messageViewHolder.itemView.context,
                                    ImageViewerActivity::class.java
                                )
                                intent.putExtra(
                                    "url",
                                    userMessagesList[position].message
                                )
                                messageViewHolder.itemView.context.startActivity(intent)
                            }
                            3 -> {
                                deleteMessageForEveryOne(position, messageViewHolder)
                            }
                        }
                    }
                    builder.show()
                }
            }
        } else {
            messageViewHolder.itemView.setOnClickListener {
                if (userMessagesList[position]
                        .type == "pdf" || userMessagesList[position]
                        .type == "docx"
                ) {
                    val options =
                        arrayOf<CharSequence>(
                            "Delete For me",
                            "Download and View this Document",
                            "Cancel"
                        )
                    val builder =
                        AlertDialog.Builder(messageViewHolder.itemView.context)
                    builder.setTitle("Delete Message")
                    builder.setItems(options) { dialogInterface, position ->
                        if (position == 0) {
                            deleteReceiverMessage(position, messageViewHolder)
                            val intent = Intent(
                                messageViewHolder.itemView.context,
                                LandingActivity::class.java
                            )
                            messageViewHolder.itemView.context.startActivity(intent)
                        } else if (position == 1) {
                            val intent = Intent(
                                Intent.ACTION_VIEW,
                                Uri.parse(
                                    userMessagesList[position].message
                                )
                            )
                            messageViewHolder.itemView.context.startActivity(intent)
                        }
                    }
                    builder.show()
                } else if (userMessagesList[position].type == "text") {
                    val options =
                        arrayOf<CharSequence>(
                            "Delete For me",
                            "Cancel"
                        )
                    val builder =
                        AlertDialog.Builder(messageViewHolder.itemView.context)
                    builder.setTitle("Delete Message")
                    builder.setItems(options) { dialogInterface, position ->
                        if (position == 0) {
                            deleteReceiverMessage(position, messageViewHolder)
                            val intent = Intent(
                                messageViewHolder.itemView.context,
                                LandingActivity::class.java
                            )
                            messageViewHolder.itemView.context.startActivity(intent)
                        }
                    }
                    builder.show()
                } else if (userMessagesList[position].type == "image") {
                    val options =
                        arrayOf<CharSequence>(
                            "Delete For me",
                            "View This Image",
                            "Cancel"
                        )
                    val builder =
                        AlertDialog.Builder(messageViewHolder.itemView.context)
                    builder.setTitle("Delete Message")
                    builder.setItems(options) { dialogInterface, position ->
                        if (position == 0) {
                            deleteReceiverMessage(position, messageViewHolder)
                            val intent = Intent(
                                messageViewHolder.itemView.context,
                                LandingActivity::class.java
                            )
                            messageViewHolder.itemView.context.startActivity(intent)
                        } else if (position == 1) {
                            val intent = Intent(
                                messageViewHolder.itemView.context,
                                ImageViewerActivity::class.java
                            )
                            intent.putExtra(
                                "url",
                                userMessagesList[position].message
                            )
                            messageViewHolder.itemView.context.startActivity(intent)
                        }
                    }
                    builder.show()
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return userMessagesList.size
    }

    private fun deleteSentMessage(position: Int, holder: MessageViewHolder) {
        val rootRef = FirebaseDatabase.getInstance().reference
        rootRef.child("Messages")
            .child(userMessagesList[position].from)
            .child(userMessagesList[position].to)
            .child(userMessagesList[position].messageID)
            .removeValue().addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    Toast.makeText(
                        holder.itemView.context,
                        "Deleted Successfully.",
                        Toast.LENGTH_SHORT
                    ).show()
                } else {
                    Toast.makeText(
                        holder.itemView.context,
                        "Error Occurred.",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
    }

    private fun deleteReceiverMessage(position: Int, holder: MessageViewHolder) {
        val rootRef = FirebaseDatabase.getInstance().reference
        rootRef.child("Messages")
            .child(userMessagesList[position].to)
            .child(userMessagesList[position].from)
            .child(userMessagesList[position].messageID)
            .removeValue().addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    Toast.makeText(
                        holder.itemView.context,
                        "Deleted Successfully.",
                        Toast.LENGTH_SHORT
                    ).show()
                } else {
                    Toast.makeText(
                        holder.itemView.context,
                        "Error Occurred.",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
    }

    private fun deleteMessageForEveryOne(position: Int, holder: MessageViewHolder) {
        val rootRef = FirebaseDatabase.getInstance().reference
        rootRef.child("Messages")
            .child(userMessagesList[position].from)
            .child(userMessagesList[position].to)
            .child(userMessagesList[position].messageID)
            .removeValue().addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    rootRef.child("Messages")
                        .child(userMessagesList[position].from)
                        .child(userMessagesList[position].to)
                        .child(userMessagesList[position].messageID)
                        .removeValue()
                        .addOnCompleteListener { task ->
                            if (task.isSuccessful) {
                                Toast.makeText(
                                    holder.itemView.context,
                                    "Error Occurred.",
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        }
                } else {
                    Toast.makeText(
                        holder.itemView.context,
                        "Error Occurred.",
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }
    }

    inner class MessageViewHolder(val binding: CustomMessagesLayoutBinding) : RecyclerView.ViewHolder(binding.root)

}